## Name: Meshank Adhia

- importing librabries

In [1]:
import sklearn.datasets      # predefined datasets can be imported using sklearn
import numpy as np           # Numpy is used here to store the data in array format.              

In [2]:
# predefined dataset named 'Breast_cancer'.
cancer_ds = sklearn.datasets.load_breast_cancer()          

In [3]:
# data will be stored in x and target will be stored in y.

X = cancer_ds.data
Y = cancer_ds.target

In [4]:
#Print variables X & Y

print(X)    
print(Y)

[[1.799e+01 1.038e+01 1.228e+02 ... 2.654e-01 4.601e-01 1.189e-01]
 [2.057e+01 1.777e+01 1.329e+02 ... 1.860e-01 2.750e-01 8.902e-02]
 [1.969e+01 2.125e+01 1.300e+02 ... 2.430e-01 3.613e-01 8.758e-02]
 ...
 [1.660e+01 2.808e+01 1.083e+02 ... 1.418e-01 2.218e-01 7.820e-02]
 [2.060e+01 2.933e+01 1.401e+02 ... 2.650e-01 4.087e-01 1.240e-01]
 [7.760e+00 2.454e+01 4.792e+01 ... 0.000e+00 2.871e-01 7.039e-02]]
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 1 0 1 0 0
 1 0 1 0 0 1 1 1 0 0 1 0 0 0 1 1 1 0 1 1 0 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1
 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1 1 0 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 0 1 1 0 0 1 1 0 0 1 1 1 1 0 1 1 0 0 0 1 0
 1 0 1 1 1 0 1 1 0 0 1 0 0 0 0 1 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 1 1 0 0 1 1
 1 0 1 1 1 1 1 0 0 1 1 0 1 1 0 0 1 0 1 1 1 1 0 1 1 1 1 1 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 1 1 1 1 0 1 0 1 1 0 1 1 0 1 0 0 1 1 1 1 1 1 1 1 

In [5]:
#shape attribute is used to count the number of rows and columns in the dataset.
#Where 0th index shows the number of rows and 1st index shows number of columns in tuple.

print(X.shape, Y.shape)                                             

(569, 30) (569,)


In [6]:
#converting into dataframe
#Using pandas library array can be converted into dataframe.

import pandas as pd       
data = pd.DataFrame(cancer_ds.data,columns = cancer_ds.feature_names)

In [7]:
#Target variable stored in columned named class 

data['class'] = cancer_ds.target

In [8]:
#head() function is used to display the First few tuples(Rows) in the dataset.
#Bydefault it will print first five rows.

data.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  class  
0          0.4601                  0.11890      0  
1          0.2750                  0.08902      0  
2          0.3613                  0.08758      0  
3          0.6638                  0.17300      0  
4          0.2364                  0.07678      0  

[5 rows x 31 columns]

In [9]:
#describe() function is used to view some basic statistical details like count,mean,std etc. of a dataframe.
data.describe()            

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst texture  \
count     569.000000              569.000000  ...     569.000000   
mean        0.181162                0.062798  ...      25.677223   
std         0.027414                0.007060  ...       6.146258   
min         0.106000                0.049960  ...      12.020000   
25%         0.161900                0.057700  ...      21.080000   
50%         0.179200                0.061540  ...      25.410000   
75%         0.195700                0.066120  ...      29.720000   
max         0.304000                0.097440  ...      49.540000   

       worst perimeter   worst area  worst smoothness  worst compactness  \
count       569.000000   569.000000        569.000000         569.000000   
mean        107.261213   880.583128          0.132369           0.254265   
std          33.602542   569.356993          0.022832           0.157336   
min          50.410000   185.200000          0.071170           0.027290   
25%          84.110000   515.300000          0.116600           0.147200   
50%          97.660000   686.500000          0.131300           0.211900   
75%         125.400000  1084.000000          0.146000           0.339100   
max         251.200000  4254.000000          0.222600           1.058000   

       worst concavity  worst concave points  worst symmetry  \
count       569.000000            569.000000      569.000000   
mean          0.272188              0.114606        0.290076   
std           0.208624              0.065732        0.061867   
min           0.000000              0.000000        0.156500   
25%           0.114500              0.064930        0.250400   
50%           0.226700              0.099930        0.282200   
75%           0.382900              0.161400        0.317900   
max           1.252000              0.291000        0.663800   

       worst fractal dimension       class  
count               569.000000  569.000000  
mean                  0.083946    0.627417  
std                   0.018061    0.483918  
min                   0.055040    0.000000  
25%                   0.071460    0.000000  
50%                   0.080040    1.000000  
75%                   0.092080    1.000000  
max                   0.207500    1.000000  

[8 rows x 31 columns]

In [10]:
#How many malignant and benign ? OR (1s and 0s)?
#Binary Variable
#value_counts() function is used to count the number of occurances.

data['class'].value_counts() 

1    357
0    212
Name: class, dtype: int64

In [11]:
#know the target names

cancer_ds.target_names

array(['malignant', 'benign'], dtype='<U9')

In [12]:
# groupby() function is used to create a group based on parameter passed in to groupby function.
# Here class is passed as an argument.

data.groupby('class').mean()    

mean radius  mean texture  mean perimeter   mean area  mean smoothness  \
class                                                                           
0        17.462830     21.604906      115.365377  978.376415         0.102898   
1        12.146524     17.914762       78.075406  462.790196         0.092478   

       mean compactness  mean concavity  mean concave points  mean symmetry  \
class                                                                         
0              0.145188        0.160775             0.087990       0.192909   
1              0.080085        0.046058             0.025717       0.174186   

       mean fractal dimension  ...  worst radius  worst texture  \
class                          ...                                
0                    0.062680  ...     21.134811      29.318208   
1                    0.062867  ...     13.379801      23.515070   

       worst perimeter   worst area  worst smoothness  worst compactness  \
class                                                                      
0           141.370330  1422.286321          0.144845           0.374824   
1            87.005938   558.899440          0.124959           0.182673   

       worst concavity  worst concave points  worst symmetry  \
class                                                          
0             0.450606              0.182237        0.323468   
1             0.166238              0.074444        0.270246   

       worst fractal dimension  
class                           
0                     0.091530  
1                     0.079442  

[2 rows x 30 columns]

In [13]:
# mean() function is used to count the average.

data.groupby('mean texture').mean()     

mean radius  mean perimeter  mean area  mean smoothness  \
mean texture                                                            
9.71                10.80           68.77      357.6          0.09594   
10.38               17.99          122.80     1001.0          0.11840   
10.72               11.54           73.73      409.1          0.08597   
10.82               11.41           73.34      403.3          0.09373   
10.89               14.50           94.28      640.7          0.11010   
...                   ...             ...        ...              ...   
31.12               20.73          135.70     1419.0          0.09469   
32.47               19.53          128.00     1223.0          0.08420   
33.56               15.53          103.70      744.9          0.10630   
33.81               11.22           70.79      386.8          0.07780   
39.28               17.46          113.40      920.6          0.09812   

              mean compactness  mean concavity  mean concave points  \
mean texture                                                          
9.71                   0.05736        0.025310             0.016980   
10.38                  0.27760        0.300100             0.147100   
10.72                  0.05969        0.013670             0.008907   
10.82                  0.06685        0.035120             0.026230   
10.89                  0.10990        0.088420             0.057780   
...                        ...             ...                  ...   
31.12                  0.11430        0.136700             0.086460   
32.47                  0.11300        0.114500             0.066370   
33.56                  0.16390        0.175100             0.083990   
33.81                  0.03574        0.004967             0.006434   
39.28                  0.12980        0.141700             0.088110   

              mean symmetry  mean fractal dimension  radius error  ...  \
mean texture                                                       ...   
9.71                 0.1381                 0.06400        0.1728  ...   
10.38                0.2419                 0.07871        1.0950  ...   
10.72                0.1833                 0.06100        0.1312  ...   
10.82                0.1667                 0.06113        0.1408  ...   
10.89                0.1856                 0.06402        0.2929  ...   
...                     ...                     ...           ...  ...   
31.12                0.1769                 0.05674        1.1720  ...   
32.47                0.1428                 0.05313        0.7392  ...   
33.56                0.2091                 0.06650        0.2419  ...   
33.81                0.1845                 0.05828        0.2239  ...   
39.28                0.1809                 0.05966        0.5366  ...   

              worst texture  worst perimeter  worst area  worst smoothness  \
mean texture                                                                 
9.71                  12.02            73.66       414.0           0.14360   
10.38                 17.33           184.60      2019.0           0.16220   
10.72                 12.87            81.23       467.8           0.10920   
10.82                 15.97            83.74       510.5           0.15480   
10.89                 15.98           102.80       745.5           0.13130   
...                     ...              ...         ...               ...   
31.12                 47.16           214.00      3432.0           0.14010   
32.47                 45.41           180.20      2477.0           0.14080   
33.56                 49.54           126.30      1035.0           0.18830   
33.81                 41.78            78.44       470.9           0.09994   
39.28                 44.87           141.20      1408.0           0.13650   

              worst compactness  worst concavity  worst concave points  \
mean texture                                                             
9.71          

# Train Test Split

In [14]:
#For spliting the dataset import the library from sklearn model_selection package.

from sklearn.model_selection import train_test_split   

In [15]:
#axis=1 for columns ::: axis=0 for rows
X = data.drop('class', axis=1) 
Y = data['class']

In [16]:
#It will split the dataset into training and testing dataset.
#Test_size=0.1 so the size of test dataset=10% & train dataset=90%

#X_train, X_test, Y_train, Y_test = train_test_split(X,Y)      

X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.1)

In [17]:
print(Y.shape, Y_train.shape, Y_test.shape)   

(569,) (512,) (57,)


In [18]:
#stratification:The train_test_split method returns training and test subsets that have the same proportions of class labels 
#as the input dataset.

X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.1, stratify = Y)
print(Y.shape, Y_train.shape, Y_test.shape)

(569,) (512,) (57,)


In [19]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.2) 
print(Y.shape, Y_train.shape, Y_test.shape)

(569,) (455,) (114,)


In [20]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.2, stratify= Y)
print(Y.shape, Y_train.shape, Y_test.shape)

(569,) (455,) (114,)


In [21]:
print(Y.mean(), Y_train.mean(), Y_test.mean())  

0.6274165202108963 0.6263736263736264 0.631578947368421


In [22]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.2, stratify= Y, random_state=2)
print(Y.mean(), Y_train.mean(), Y_test.mean())

0.6274165202108963 0.6263736263736264 0.631578947368421


In [23]:
print(X.mean(), X_train.mean(), X_test.mean())

mean radius                 14.127292
mean texture                19.289649
mean perimeter              91.969033
mean area                  654.889104
mean smoothness              0.096360
mean compactness             0.104341
mean concavity               0.088799
mean concave points          0.048919
mean symmetry                0.181162
mean fractal dimension       0.062798
radius error                 0.405172
texture error                1.216853
perimeter error              2.866059
area error                  40.337079
smoothness error             0.007041
compactness error            0.025478
concavity error              0.031894
concave points error         0.011796
symmetry error               0.020542
fractal dimension error      0.003795
worst radius                16.269190
worst texture               25.677223
worst perimeter            107.261213
worst area                 880.583128
worst smoothness             0.132369
worst compactness            0.254265
worst concav

In [24]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.2, stratify= Y, random_state=2)
print(Y.mean(), Y_train.mean(), Y_test.mean())
print(X.mean(), X_train.mean(), X_test.mean())

0.6274165202108963 0.6263736263736264 0.631578947368421
mean radius                 14.127292
mean texture                19.289649
mean perimeter              91.969033
mean area                  654.889104
mean smoothness              0.096360
mean compactness             0.104341
mean concavity               0.088799
mean concave points          0.048919
mean symmetry                0.181162
mean fractal dimension       0.062798
radius error                 0.405172
texture error                1.216853
perimeter error              2.866059
area error                  40.337079
smoothness error             0.007041
compactness error            0.025478
concavity error              0.031894
concave points error         0.011796
symmetry error               0.020542
fractal dimension error      0.003795
worst radius                16.269190
worst texture               25.677223
worst perimeter            107.261213
worst area                 880.583128
worst smoothness             0.1

In [25]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.2, random_state=124)
print(Y.mean(), Y_train.mean(), Y_test.mean())
print(X.mean().head(), X_train.mean().head(), X_test.mean().head())

0.6274165202108963 0.6373626373626373 0.5877192982456141
mean radius         14.127292
mean texture        19.289649
mean perimeter      91.969033
mean area          654.889104
mean smoothness      0.096360
dtype: float64 mean radius         14.066582
mean texture        19.347297
mean perimeter      91.559626
mean area          649.874945
mean smoothness      0.096429
dtype: float64 mean radius         14.369596
mean texture        19.059561
mean perimeter      93.603070
mean area          674.901754
mean smoothness      0.096085
dtype: float64


In [26]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.2, random_state=999)
print(Y.mean(), Y_train.mean(), Y_test.mean())
print(X.mean().head(), X_train.mean().head(), X_test.mean().head())

0.6274165202108963 0.6461538461538462 0.5526315789473685
mean radius         14.127292
mean texture        19.289649
mean perimeter      91.969033
mean area          654.889104
mean smoothness      0.096360
dtype: float64 mean radius         14.031699
mean texture        19.143121
mean perimeter      91.305253
mean area          644.758462
mean smoothness      0.095914
dtype: float64 mean radius         14.508825
mean texture        19.874474
mean perimeter      94.618333
mean area          695.322807
mean smoothness      0.098140
dtype: float64


In [27]:
#Binarization of Input
X_binarised_train = X_train.apply(pd.cut, bins=2, labels =[1,0])
X_binarised_test = X_test.apply(pd.cut, bins=2, labels =[1,0])
type(X_binarised_train)
X_binarised_train = X_binarised_train.values
X_binarised_test = X_binarised_test.values
type(X_binarised_train)

numpy.ndarray

In [28]:
X_binarised_train.shape

(455, 30)

In [29]:
#what are the possible values that b can take? 

b=10
i = 100
if(np.sum(X_binarised_train[i,:])>=b):
    print('Model Prediction is Malignant')
else:
    print('Model prediction is Benign')
    
if(Y_train[i]==1):
    print('Actual Outcome is Malignant')
else:
    print('Actual Outcome is Benign')

Model Prediction is Malignant
Actual Outcome is Benign


In [30]:
b=10

from random import randint
i=randint(0,X_binarised_train.shape[0])

if(np.sum(X_binarised_train[i,:])>=b):
    print('Model Prediction is Malignant')
else:
    print('Model prediction is Benign')
    
if(Y_train[i]==1):
    print('Actual Outcome is Malignant')
else:
    print('Actual Outcome is Benign')

Model Prediction is Malignant
Actual Outcome is Malignant


In [31]:
b=10

from random import randint
#i=randint(0,X_binarised_train.shape[0])

accurate_rows= 0
for x,y in zip(X_binarised_train,Y_train):
    y_pred=(np.sum(x)>=b)
    accurate_rows += (y == y_pred)
    
print(accurate_rows, accurate_rows/X_binarised_train.shape[0])


294 0.6461538461538462


In [32]:
#Accuracy = number of rows predicted correctly/total number of rows
#from random import randint
#i=randint(0,X_binarised_train.shape[0])
i=100
for b in range(X_binarised_train.shape[1]+1):
    accurate_rows= 0
    for x,y in zip(X_binarised_train,Y_train):
        y_pred=(np.sum(x)>=b)
        accurate_rows += (y == y_pred)
    print("For b = ",b,accurate_rows, accurate_rows/X_binarised_train.shape[0])

For b =  0 294 0.6461538461538462
For b =  1 294 0.6461538461538462
For b =  2 294 0.6461538461538462
For b =  3 294 0.6461538461538462
For b =  4 294 0.6461538461538462
For b =  5 294 0.6461538461538462
For b =  6 294 0.6461538461538462
For b =  7 294 0.6461538461538462
For b =  8 294 0.6461538461538462
For b =  9 294 0.6461538461538462
For b =  10 294 0.6461538461538462
For b =  11 294 0.6461538461538462
For b =  12 294 0.6461538461538462
For b =  13 296 0.6505494505494506
For b =  14 298 0.654945054945055
For b =  15 299 0.6571428571428571
For b =  16 300 0.6593406593406593
For b =  17 303 0.6659340659340659
For b =  18 306 0.6725274725274726
For b =  19 309 0.6791208791208792
For b =  20 311 0.6835164835164835
For b =  21 320 0.7032967032967034
For b =  22 329 0.7230769230769231
For b =  23 340 0.7472527472527473
For b =  24 355 0.7802197802197802
For b =  25 368 0.8087912087912088
For b =  26 376 0.8263736263736263
For b =  27 383 0.8417582417582418
For b =  28 388 0.8527472527472

In [33]:
b=28
i=100
print("b =",b)
for b in range(X_binarised_test.shape[1]+1):
    accurate_rows= 0
    for x,y in zip(X_binarised_train,Y_train):
        y_pred=(np.sum(x)>=b)
        accurate_rows += (y == y_pred)
print(accurate_rows, accurate_rows/X_binarised_train.shape[0])


b=27
i=100
print("b =",b)
for b in range(X_binarised_test.shape[1]+1):
    accurate_rows= 0
    for x,y in zip(X_binarised_train,Y_train):
        y_pred=(np.sum(x)>=b)
        accurate_rows += (y == y_pred)
print(accurate_rows, accurate_rows/X_binarised_train.shape[0])

b = 28
338 0.7428571428571429
b = 27
338 0.7428571428571429


# Conclusion: 
## By This Breast cancer dataset can be analysed & Different Machine learnign classification model can be Generated to compare the results & mostly best model is identified based on the nural network. By using inbuilt function,predicted outcome & actual outcome can be classified into 'malignant' OR 'benign'.
## Firstly compilation of the model is done and then fit the model.
## Accuracy of whole model is 74.29%